hack/util.sh

```diff
@@ -44,50 +75,132 @@ nodes:
+
+  rm "${KWOK_CONFIG}"
+}
+
 function util::create_member_cluster() {
   local MEMBER_CLUSTER_NAME=${1}
   local KUBECONFIG_PATH=${2}
 
-  kind create cluster --image=kindest/node:v1.20.15 --name="${MEMBER_CLUSTER_NAME}" --kubeconfig="${KUBECONFIG_PATH}"
+  if [[ $CLUSTER_PROVIDER == "kind" ]]; then
+    kind create cluster --image=kindest/node:v1.20.15 --name="${MEMBER_CLUSTER_NAME}" --kubeconfig="${KUBECONFIG_PATH}"
+  elif [[ $CLUSTER_PROVIDER == "kwok" ]]; then
+    APISERVER_PORT=$(shuf -i 30000-40000 -n 1)
+    KUBECONFIG=${KUBECONFIG_PATH} KWOK_KUBE_VERSION="v1.20.15" kwokctl create cluster --name=${MEMBER_CLUSTER_NAME} --kube-apiserver-port=${APISERVER_PORT} --kube-authorization
+    for i in $(seq 1 10); do
+      kubectl --kubeconfig="${KUBECONFIG_PATH}" --context="kwok-${MEMBER_CLUSTER_NAME}" apply -f - <<EOF
+apiVersion: v1
+kind: Node
+metadata:
+  annotations:
+    node.alpha.kubernetes.io/ttl: "0"
+    kwok.x-k8s.io/node: fake
+  labels:
+    beta.kubernetes.io/arch: amd64
+    beta.kubernetes.io/os: linux
+    kubernetes.io/arch: amd64
+    kubernetes.io/hostname: kwok-node-0
+    kubernetes.io/os: linux
+    kubernetes.io/role: agent
+    node-role.kubernetes.io/agent: ""
+    type: kwok
+  name: kwok-node-${i}
+status:
+  allocatable:
+    cpu: 32
+    memory: 256Gi
+    pods: 110
+  capacity:
+    cpu: 32
+    memory: 256Gi
+    pods: 110
+  nodeInfo:
+    architecture: amd64
+    bootID: ""
+    containerRuntimeVersion: ""
+    kernelVersion: ""
+    kubeProxyVersion: fake
+    kubeletVersion: fake
+    machineID: ""
+    operatingSystem: linux
+    osImage: ""
+    systemUUID: ""
+  phase: Running
+EOF
+    done
+  else
+    echo "Invalid provider, only kwok or kind allowed"
+    exit 1
+  fi
 }
 
 function util::join_member_cluster() {
   local MEMBER_CLUSTER_NAME=${1}
   local HOST_CLUSTER_NAME=${2}
   local KUBECONFIG_PATH=${3}
 
-  local HOST_CLUSTER_KUBECONFIG
-  HOST_CLUSTER_KUBECONFIG=$(mktemp)
-  kind get kubeconfig --name="${HOST_CLUSTER_NAME}" > "${HOST_CLUSTER_KUBECONFIG}"
-
-  local MEMBER_API_ENDPOINT MEMBER_CA MEMBER_CERT MEMBER_KEY
-  MEMBER_API_ENDPOINT=$(kind get kubeconfig --name="${MEMBER_CLUSTER_NAME}" | grep 'server:' | awk '{ print $2 }')
-  MEMBER_CA=$(kind get kubeconfig --name="${MEMBER_CLUSTER_NAME}" | grep 'certificate-authority-data:' | awk '{ print $2 }')
-  MEMBER_CERT=$(kind get kubeconfig --name="${MEMBER_CLUSTER_NAME}" | grep 'client-certificate-data:' | awk '{ print $2 }')
-  MEMBER_KEY=$(kind get kubeconfig --name="${MEMBER_CLUSTER_NAME}" | grep 'client-key-data:' | awk '{ print $2 }')
-
-  local FCLUSTER FCLUSTER_SECRET
-  FCLUSTER=$(mktemp)
-  FCLUSTER_SECRET=$(mktemp)
-
-  cat <<EOF > "${FCLUSTER}"
+  local HOST_CONTEXT MEMBER_API_ENDPOINT MEMBER_CA MEMBER_CERT MEMBER_KEY
+  if [[ $CLUSTER_PROVIDER == "kind" ]]; then
+    HOST_CONTEXT="kind-${HOST_CLUSTER_NAME}"
+    MEMBER_API_ENDPOINT=$(kind get kubeconfig --name="${MEMBER_CLUSTER_NAME}" | grep 'server:' | awk '{ print $2 }')
```

Comment:
Why don't we get these from the kubeconfig directly?

Reply:
We wouldn't need to maintain 2 sets of logic if getting from kubeconfig directly. But not a big issue.